### PR TITLE
LPS-107816

### PR DESCRIPTION
**Pull request by jesseyeh-liferay**
**[LPS-107816](https://issues.liferay.com/browse/LPS-107816)**

**Previous pull requests**: https://github.com/achaparro/liferay-portal/pull/894 > https://github.com/achaparro/liferay-portal/pull/896 > https://github.com/achaparro/liferay-portal/pull/902

> This changes the type of the version column in KaleoDefinitionVersion from a String to an int.

**Changelog**:

**Upgrading from 6.2**
In the current implementation in `master`, `KaleoDefinitionVersion` only accounts for entries from the `KaleoDefinition` table. In order to include the missing draft versions, [additional entries need to be processed from the `KaleoDraftDefinition` table as well](https://github.com/achaparro/liferay-portal/blob/a62954d024e1f80c21f71e478deec540da9e4013/modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v1_4_1/KaleoDefinitionVersionUpgradeProcess.java#L197-L198).

**Handling Decimal-Style Versions**
This addresses [an issue](https://github.com/liferay-workflow/liferay-portal/pull/375#issuecomment-848197114) in which `KaleoDefinitionVersion.version` is stored in a decimal-style format, e.g., "1.1", "2.3", etc. To resolve this, [all of the entries are first sorted by their name, then by the length of their versions, and then by their decimal value in ascending order](https://github.com/achaparro/liferay-portal/blob/a62954d024e1f80c21f71e478deec540da9e4013/modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v4_0_0/UpgradeKaleoDefinitionVersion.java#L37-L38). Doing so prevents a situation in which, for example, "2.10" (length of 4) would appear before "2.2" (length of 3).

The sorted `KaleoDefinitionVersion` entries are then remapped to a new version number and the corresponding entries in `KaleoDefinition` are updated. Example:
|Before|After|
|---|---|
|1.0|1|
|1.1|2|
|1.2|3|
|2.0|4|

This solution makes the following assumptions:
1. Entries in `KaleoDefinitionVersion` can have either integer-style or decimal-style versions, but never both at the same time. For example, there cannot be an entry with version "1" and another entry with version "1.3." There can, however, be an entry with version "1.0" and another entry with version "1.3."
2. The largest possible version number is "99.99" (see `cast(version as decimal(4, 2))` in the `ORDER BY` clause of the sort)

**Known Issue**
After upgrading, entries in the `KaleoDefinition` table are all consolidated into a single entry. For example, continuing off of the previous before/after table:
|Before|After (Expected)|After (Actual)|
|---|---|---|
|KD entry version 1<br/>KD entry version 2|KD entry version 1<br/>KD entry version 4|KD entry version 4|

### CODE REVIEW

**Comment by jesseyeh-liferay** [comment]
ci:test:sf

**Comment by jesseyeh-liferay** [comment]
ci:test:relevant

**Comment by jesseyeh-liferay** [comment]
ci:stop

**Comment by jesseyeh-liferay** [comment]
ci:test:sf

**Comment by jesseyeh-liferay** [comment]
ci:test:relevant

**Comment by jesseyeh-liferay** [comment]
ci:stop

**Comment by jesseyeh-liferay** [comment]
ci:test:sf

**Comment by jesseyeh-liferay** [comment]
ci:test:relevant

**Comment by jesseyeh-liferay** [comment]
ci:test:sf

**Comment by jesseyeh-liferay** [comment]
ci:test:relevant

**Comment by jesseyeh-liferay** [comment]
ci:test:sf

**Comment by jesseyeh-liferay** [comment]
ci:test:relevant

**Comment by jesseyeh-liferay** [comment]
ci:stop

**Comment by jesseyeh-liferay** [comment]
ci:test:sf

**Comment by jesseyeh-liferay** [comment]
ci:test:relevant

**Comment by achaparro** [comment]
Just started reviewing :)

File: modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v1_4_1/KaleoDefinitionVersionUpgradeProcess.java [hunk 1]
```diff
@@ -186,6 +196,62 @@ protected void upgradeKaleoDefinitionVersion() throws Exception {
...
+			DBInspector dbInspector = new DBInspector(connection);
+
+			DatabaseMetaData metaData = connection.getMetaData();
+
+			ResultSet metaDataResultSet = metaData.getTables(
+				dbInspector.getCatalog(), dbInspector.getSchema(),
+				"KaleoDraftDefinition", new String[] {"TABLE"});
+
+			if (metaDataResultSet.next()) {
+				PreparedStatement preparedStatement3 =
```
Comment (achaparro) [inline modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v1_4_1/KaleoDefinitionVersionUpgradeProcess.java:208]: @jesseyeh-liferay does the upgrade _com.liferay.portal.workflow.kaleo.designer.web.internal.upgrade.v1_0_1.KaleoDefinitionVersionUpgradeProcess_ make sense after adding this code here?

Comment (jesseyeh-liferay) [reply modules/apps/portal-workflow/portal-workflow-kaleo-service/src/main/java/com/liferay/portal/workflow/kaleo/internal/upgrade/v1_4_1/KaleoDefinitionVersionUpgradeProcess.java:208]: @achaparro 
I tested this by using a dummy upgrade step and found that the upgraded KaleoDefinitionVersion table was identical. See https://github.com/achaparro/liferay-portal/pull/937/commits/0d838e469b120db2b6a4249171389b641507a76a.

**Comment by achaparro** [comment]
Just started reviewing :)

**Comment by achaparro** [comment]
Pull request submitted to https://github.com/jesseyeh-liferay/liferay-portal/pull/29. [See changes here.](https://github.com/achaparro/liferay-portal/compare/2286180ff191~2...2286180ff191)

**Pull request closed without merging**